proposals/0000-hybrid-output.md

```diff
@@ -0,0 +1,133 @@
+
+# Detailed design
+
+## Public API
+
+There are two public API surfaces to this RFC:
+
+- Updating the `astro.config.mjs` to accept a new `output` value, `'hybrid'`. The existing values of `'static' | 'server'` would remain supported. A value of `'hybrid'` will enabled detection for the following feature:
+- Allowing routes (files in `pages/`) to self-declare a granular `output` value. The existing `output` values of `'static' | 'server'` would both be valid.
+
+**For performance reasons, this declaration must be statically analyzable.**
+Only string literal values will be accepted. Other declaration values will throw an error and fail the build. This mirrors similar constraints that Vite has for compile-time (macro) features like dynamic `import()` and `import.meta.glob`.
+
+```astro
+---
+// Valid – constant string literal
+export const output = 'server'
+// Invalid – cannot be determined at compile-time
+export const output = static ? 'static' : 'server';
+export const output = value;
+export const output = 'ser' + 'ver';
+export let output = undefined; output = 'server';
+---
+```
+
+## Internal API
+
+Astro needs to detect/understand the `export const output` syntax. This should be implemented as a specific post-processing Babel plugin. The plugin will detect the existence of valid ESM exports of `output` (`export const output`, `export let output`, `export var output`, `export { output }`, etc) and attach the statically detected `output` value to the module metadata under the `astro` namespace. This will allow us to track the `output` mode of any public route.
+
+> **Note**
+> Astro Routes include `.astro` pages, but also API endpoints (`src/pages/api/user.ts`)! It's important that this RFC handles both of these constraints in a uniform way, hence the use of a post-processing Babel plugin rather than implementing this feature in `@astrojs/compiler`.
+
+The **build** API needs to be updated to handle the new `'hybrid'` output mode. Implementation-wise, this will be similar to the existing `'server'` implementation, but it will also adopt portions of the `'static'` build to statically generate any routes that are not handled at request time.
+
+The **dev** server must be updated to allow `searchParams` and other features which are currently only avaliable with `server` rendering to be allowed on a per-route basis.
+
+## Adapter Support
+
+An adapter will be required when building for `hybrid` output, because we will be building a server with some statically generated routes. Adapters will need to be updated to handle `hybrid` output.
+
+# Testing strategy
+
+This implementation will largely be tested with full fixture-based integration tests, because we will need to verify the build output is structured correctly, which will be difficult to do with unit tests.
+
+The Babel plugin that detects `export const output` will be unit tested against all valid `export` formats outlined above.
```

Comment:
Will it be tested against the _invalid_ formats too?

Reply:
Yep, updated!